### DISCUSSION TITLE
PDF issues in table and table of content

### DISCUSSION THREAD

**Discussion by sachin-suresh-rapyuta** [brechtm/rinohtype#337] [Q&A]
Hi, when I generate the PDF, I see the following issues:

1. Table breaks even though I have specified `tabularcolumns` directive.
2. I don't see my TOC displayed in PDF

**Issue 1:** Table breaks although it appears fine in the HTML

Here is the code:

```
.. tabularcolumns:: |\Yl{0.25}|\Yl{0.25}|\Yl{0.25}|

      .. list-table:: List of available request types
         :widths: 25 25 25
         :header-rows: 1
         :class: tight-table

         * - Name
           - Class (foundation_model::...) 
           - Description
         * - IDLE
           - ``IdleRequest``
           - The robot waits
         * - MOVE_TO
           - ``MoveToRequest``
           - Send the robot to a specific location
         * - ORDER
           - ``OrderRequest``
           - Complete an OrderRequest (1 Tote)
```

Command used to generate the PDF: `sphinx-build -b rinoh source build/rinoh`

*Issue 2** TOC not displayed

I have the same code added as per this thread for getting TOC- https://github.com/brechtm/lizmap-documentation/commit/2c6d5cfb1a023131fab4fd65bbb159ecc4175caf#commitcomment-76624408

But, I dont see the TOC in my PDF.

Version details
------------------------
sphinx-build 3.2.1
Python 3.8.10
rinohtype 0.5.4

**Comment by brechtm** [comment]
### Issue 1 - tabular columns

`tabularcolumns` is only needed when using the LaTeX builder. rinohtype features automatic table column sizing. Not specifying `:widths:` will usually provide good results.

That said, rinohtype simply ignores the `tabularcolumns`. However, I see that you have placed the `list-table` in the body of the `tabularcolumns` directive. Note that the Sphinx manual says (emphasis mine)

> This directive gives a “column spec” for the **next** table occurring in the source file.

So, you actually want this:

```rst
.. tabularcolumns:: |\Yl{0.25}|\Yl{0.25}|\Yl{0.25}|

.. list-table:: List of available request types
     :header-rows: 1
     :class: tight-table

     * - Name
       - Class (foundation_model::...)
       - Description
     * - IDLE
       - ``IdleRequest``
       - The robot waits
     * - MOVE_TO
       - ``MoveToRequest``
       - Send the robot to a specific location
     * - ORDER
       - ``OrderRequest``
       - Complete an OrderRequest (1 Tote)
```

But, again, if you're not using the LaTeX builder, you can simply omit the `tabularcolumns` directive.

### Issue 2 - missing TOC

Please attach your Sphinx project. I have nothing to go by otherwise, I'm afraid.

**Comment by sachin-suresh-rapyuta** [comment]
I am attaching my test [Test.zip](https://github.com/brechtm/rinohtype/files/8959051/Test.zip) sphinx project.

Few [observations:]([Test.zip](https://github.com/brechtm/rinohtype/files/8959046/Test.zip))
1. In **testfile1.rst**, the table code you have given works in PDF. But I have used similar code on top of that - 'Warehouse params'. But this has 4 columns for which the table still breaks in PDF.
2. See the missing TOC or a blank page is appearing before the page 1/5. On this blank page is where I want my Title (see point 3) followed by Table of Contents.
3. The Tiltle from **index.rst**, "Welcome to Test project's documentation!" is not appearing in PDF.

**Reply by sachin-suresh-rapyuta** [reply]
It would also be great if you can provide us a sample stylesheet. rts which has style codes for some beautified PDFs. The [example](http://www.mos6581.org/rinohtype/master/basicstyling.html#extending-an-existing-style-sheet) given is somewhat limited?

**Reply by sachin-suresh-rapyuta** [reply]
@brechtm  were there any updates on this?

**Reply by brechtm** [reply]
I haven't found the time yet to look into this. Hopefully somewhere this week 🤞

**Reply by brechtm** [reply]
Sorry this took so long!

> 1. In **testfile1.rst**, the table code you have given works in PDF. But I have used similar code on top of that - 'Warehouse params'. But this has 4 columns for which the table still breaks in PDF.

In this case, there is simply not enough room to fit all four columns into the available width. Even if you were to give the first two columns enough room to fit their contents, almost nothing is left for the paragraphs in the fourth column.

Some options you have:
- Choose a larger paper size
- Decrease the font size. Add this to the _my_stylesheet.rts_ style sheet.
    ```
    [default:Paragraph]
    font_size = 8pt
    ```

- Decrease the page margins. Add the following to _rinohtype.rtt_:
    ```
    [contents_page]
    left_margin = 1cm
    right_margin = 1cm
    ```

A [definition list](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#definition-lists) is probably a better format for presenting this information.

> 2. See the missing TOC or a blank page is appearing before the page 1/5. On this blank page is where I want my Title (see point 3) followed by Table of Contents.

The table of contents is part of the front matter; you need to uncomment the `front_matter` line in the _rinohtype.rtt_ template configuration file.
 
> 3. The Tiltle from **index.rst**, "Welcome to Test project's documentation!" is not appearing in PDF.

Let's discuss this one in #338. Please address my questions there.

**Reply by brechtm** [reply]
> It would also be great if you can provide us a sample stylesheet. rts which has style codes for some beautified PDFs. The [example](http://www.mos6581.org/rinohtype/master/basicstyling.html#extending-an-existing-style-sheet) given is somewhat limited?

Yes, that is something that is definitely missing from the documentation. The idea is that a selection of properly designed templates/stylesheets will be available to users, but that hasn't materialized yet.
